Bug 1226652 - Login List Multiselection and Deletion

sleroux opened the pull request:
* Added multi selection functionality to login list
* Search view now retains search after navigating to detail view
* Selected logins from list can be deleted

rnewman commented on Client/Extensions/UIAlertControllerExtensions.swift:
```diff
@@ -118,4 +118,49 @@ extension UIAlertController {
+     case of local-only logins or deleted across synced devices in synced account logins.
+
+     - parameter deleteCallback: Block to run when delete is tapped.
+     - parameter hasAccount: Boolean indicating the user has a FxA account connected.
```
As with #1370, this isn't correct — if you're not syncing passwords, you're going to be really alarmed to be told that the logins you delete will be removed from your other devices.

(Also please steal that review request from @st3fan!)

sleroux replied:
So on desktop, I can configure what I want to sync so it's not enough to know if they have an account, I also need to know if they've enabled password syncing. 

Looking through the code it seems this configuration is part of the 'sync' prefs branch:

https://github.com/mozilla/firefox-ios/blob/master/Providers/Profile.swift#L685

Looks like if metaGlobal scratchpad doesn't include the 'passwords' collection, that means that password sync is not enabled (inferred from: https://github.com/mozilla/firefox-ios/blob/master/Sync/Synchronizers/Synchronizer.swift#L177) and this scratchpad lives inside the SyncStateMachine.

What would be the best way to expose this?

rnewman replied:
The simplest — and most accurate — thing to do is to ask `SyncablePasswords` if there are any passwords that would propagate as deletions; not only does this avoid trying to extract state from the account, which is complicated and error-prone, but it also correctly muffles the warning if they haven't synced passwords yet, have none saved, or already wiped them. Storage is the source of truth!

For history, #1370 calls this `hasSyncedHistory`, so I'd ape that.

This basically comes down to checking if `loginsM` is non-empty or `loginsL` has anything with a `sync_status` that's not `New`. We're effectively trying to figure out if `SQLiteLogins.removeAll` will leave deletion sentinel rows in `loginsL`.

rnewman commented on Client/Frontend/Login Management/LoginListViewController.swift:
```diff
@@ -70,18 +109,121 @@ class LoginListViewController: UIViewController {
+
+            let now = NSDate.now()
+            let deleteDeferreds = loginsToDelete.map { login in
+                return self.profile.logins.deleteByGUID(login.guid, deletedAt: now)
```
Nope — `deleteByGUID` is for Sync to use. Front-end code should use `removeLoginByGUID`.

Also, those deferreds will be firing immediately, and will all end up in the database queue. If you were to do this serially, do this instead:

```
walk(loginsToDelete) { self.profile.logins.deleteByGUID($0.guid, deletedAt: now }
```

which returns a single `Success`.

(And I'd suggest generalizing that to `removeLoginsWithGUIDs` so that this can be done in a single database transaction. Let me know if you need pointers on that.)

sleroux commented:
@rnewman I've updated the PR to include the removeLoginsWithGUIDs method and hasSyncableLogins check to properly message the user.

@tecgirl I've added in the the custom checkmark and fox icons as well.

rnewman commented on Client/Frontend/Login Management/LoginListViewController.swift:
```diff
@@ -119,44 +276,85 @@ extension LoginListViewController: KeyboardHelperDelegate {
+    private func searchLoginsWithText(text: String) -> Success {
+        activeLoginQuery = profile.logins.searchLoginsWithQuery(text)
+            .bindQueue(dispatch_get_main_queue(), f: reloadTableWithResult)
+        return activeLoginQuery!
```
You can drop this down to one line.

rnewman commented on Client/Frontend/Login Management/LoginListViewController.swift:
```diff
@@ -11,19 +11,54 @@ import Shared
 
     private let searchView = SearchInputView()
 
+    private var activeLoginQuery: Success?
```
I'm a little concerned that this is prone to races, but I think the impact of races won't be bad, so we're OK.

rnewman commented on Storage/MockLogins.swift:
```diff
@@ -119,11 +119,21 @@ public class MockLogins: BrowserLogins, SyncableLogins {
 
+    public func removeLoginsWithGUIDs(guids: [GUID]) -> Success {
+        return walk(guids) { guid in
+            self.removeLoginByGUID(guid)
```
File a follow-up to make this take 1 transaction not N.

rnewman commented on Storage/SQL/SQLiteLogins.swift:
```diff
@@ -502,6 +502,45 @@ public class SQLiteLogins: BrowserLogins {
             >>> effect(self.notifyLoginDidChange)
     }
 
+    public func removeLoginsWithGUIDs(guids: [GUID]) -> Success {
```
```
if guids.isEmpty {
    return succeed()
}
```

rnewman replied:
You should replace the body of `removeLoginWithGUID` with:

```
return self.removeLoginsWithGUIDs([guid])
```

rnewman commented on Storage/SQL/SQLiteLogins.swift:
```diff
@@ -890,6 +929,15 @@ extension SQLiteLogins: SyncableLogins {
+
+    public func hasSyncedLogins() -> Deferred<Maybe<Bool>> {
+        let checkLoginsMirror = "SELECT 1 FROM \(TableLoginsMirror)"
+        let checkLoginsLocal = "SELECT 1 FROM \(TableLoginsLocal) WHERE sync_status != \(SyncStatus.New.rawValue)"
```
I have a preference for `IS NOT` instead of `!=`. It works with `NULL`, and it's a little less prone to typos.

rnewman commented on Storage/SQL/SQLiteLogins.swift:
```diff
@@ -890,6 +929,15 @@ extension SQLiteLogins: SyncableLogins {
+        let checkLoginsMirror = "SELECT 1 FROM \(TableLoginsMirror)"
+        let checkLoginsLocal = "SELECT 1 FROM \(TableLoginsLocal) WHERE sync_status != \(SyncStatus.New.rawValue)"
+
+        return self.db.queryReturnsResults(checkLoginsLocal) >>== { result in
```
```
let sql = "\(checkLoginsMirror) UNION ALL \(checkLoginsLocal)"
return self.db.queryReturnsResults(sql)
```

rnewman commented:
Did I already land `queryReturnsResults`? I don't see my commits from #1370 in this stack.

sleroux commented:
Not yet - I duplicated the code from that PR.

rnewman commented:
Could you do the thirty-second test of that PR and land it?

sleroux commented:
Yup

sleroux commented:
Updated with fixes and rebased from master to include fixes from https://github.com/mozilla/firefox-ios/pull/1370

@thebnich mind taking a quick look at https://github.com/mozilla/firefox-ios/commit/35fddce040589e4b4576ca0476073b2a2fdfda4a? Had to do some hackery to accommodate the custom selection icons we want to use.

rnewman commented on Storage/SQL/SQLiteLogins.swift:
```diff
@@ -470,36 +470,50 @@ public class SQLiteLogins: BrowserLogins {
+            ", is_deleted = 1" +
+            ", password = ''" +
+            ", hostname = ''" +
+            ", username = ''" +
```
Nit: if you keep the old indentation throughout, it makes this all clearer.

rnewman commented on Storage/SQL/SQLiteLogins.swift:
```diff
@@ -470,36 +470,50 @@ public class SQLiteLogins: BrowserLogins {
-           >>> { self.db.run(insert, withArgs: args) }
-            >>> effect(self.notifyLoginDidChange)
+        let args: Args = guids.map { $0 as AnyObject }
+        return self.db.run([
```
Nice.

rnewman commented:
LGTM.

thebnich commented:
@sleroux I found one minor bug: Click **Edit**, then **Select All**, then go back, then click **Edit** again. The button says **Deselect All** even though nothing is selected.

Other than that, LGTM besides the layout question above.

sleroux commented:
Thanks for finding that bug. I've added a UITest to check for it as well as a fix.
